app/src/main/java/org/mozilla/focus/browser/BrowserFragment.kt

```diff
@@ -191,26 +190,13 @@ class BrowserFragment : IWebViewLifecycleFragment() {
     override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
         val layout = inflater.inflate(R.layout.fragment_browser, container, false)
 
-        if (isWebViewVisibleObserver != null) {
-            throw IllegalStateException("WebViewVisibleObserver unexpectedly already exists: bad lifecycle assumptions?")
-        }
-        isWebViewVisibleObserver = IsWebViewVisibleViewModelObserver().also {
-            // TODO: After SDK 28 upgrade (#72), observe over viewLifecycleOwner, remove obs removal in destroyView.
-            viewModel.isWebViewVisible.observe(this@BrowserFragment, it)
-        }
+        viewModel.isWebViewVisible.observe(viewLifecycleOwner, Observer {
+            webView?.setVisibility(if (it!!) View.VISIBLE else View.GONE)
```

Comment:
This is a minor thing, but I'm not a fan of double banging in these observers.  If a consumer ever subscribes to one before the producer emits a value, null will be passed.  Right now we're assuming that the value will always be sent before we start listening.  That's probably true most of the time, but I don't see any reason not to short on null.  It would be entirely reasonable for a `LiveData` to emit a value a few MS after `onCreateView`, rather than a few MS before.  Certainly wouldn't look any different to the user.

Reply:
> This is a minor thing, but I'm not a fan of double banging in these observers. If a consumer ever subscribes to one before the producer emits a value, null will be passed. 

I just checked and I don't see this behavior:
```kotlin
        // In Activity.onCreate...
        val test = MutableLiveData<Boolean>()
        test.observe(this, Observer {
            Log.e("lol", "haha $it") // never called
        })
```

Are there other circumstances where you expect null to be emitted?

If not, since I don't expect the producers to ever explicitly set null, it makes sense to me to 1) assert the expected state with `!!` which also 2) reduces indentation boilerplate.

In theory, we could better encapsulate one of these approaches in a `NonNullObserver`, like we have an Echo Show (originally from Focus): https://github.com/mozilla-mobile/firefox-echo-show/blob/master/app/src/main/java/org/mozilla/focus/architecture/NonNullObserver.kt But the way it's defined doesn't let you use it with a lambda so I don't really use it.

Reply:
My mistake on that one.  To be fair, it's possible to short these easily without increasing indentation.

```kotlin
        test.observe(this, Observer {
            it ?: return@Observer
            Log.e("lol", "haha $it")
        })
```

Reply:
fwiw, I think the ideal solution is to create an observer that abstracts away this decision making process.

As to which approach that abstraction takes (ignoring nulls or asserting them), for me it's a philosophical difference: if this state should _never_ happen, then I should assert it will never happen. To me, this is the same as asserting an argument passed into a function is non-null (i.e. in Java) before using it. If the assertion fails, then something terrible is wrong. In this case, e.g. one of:
- The producer incorrectly placed a null into the observer
- The observers are not attached correctly
- Something else terrible has happened

In all cases, I want to know about it and ignoring the nulls is going to hide that state from me (unless we capture these errors in Sentry and fix them when they happen, in which case, I'm less decided on which approach is best).

---

Then again, maybe we should take this one step further and create LiveData observers that are _never_ null so that this condition is protected in the type system. The `NonNullLiveData` in Echo Show actually does this (but adding NonNullObservers in Kotlin didn't use the lambda so I never used it – I really should have just fixed it). It's unclear how using `combineLatest` on NonNullLiveData will turn out too (I assume the return type is just LiveData...).

I wonder if RxJava will solve this problem at all for us. It doesn't take nulls at all, right?

Reply:
> I wonder if RxJava will solve this problem at all for us. It doesn't take nulls at all, right?

Correct.  Maybe we shouldn't worry too much about it, seeing as how we plan to switch over anyway.